### PR TITLE
Component graph and text featurizer: preserve user-specified woodwork types

### LINKED ISSUE #2296: Converting Datetime to Natural Language produces TextFeaturizer error during pipeline fit
**Reproducer**
```python
import evalml
import pandas as pd
import woodwork as ww
import pytest

df = pd.read_csv('/Users/dylan.sherry/Downloads/fraud_detection_cleaned_v2_2500rows.csv')
dt = ww.DataTable(df)
y = dt.pop('fraud')
assert str(dt.logical_types['datetime']) == 'Datetime'

# manually change datetime to natural language. strange thing to do but it is possible.
dt = dt.set_types(logical_types={'datetime': 'NaturalLanguage'})
assert str(dt.logical_types['datetime']) == 'NaturalLanguage'

# define a pipeline _without_ a datetime featurizer, since there's no datetime-typed features in the data
pipeline = evalml.pipelines.BinaryClassificationPipeline([
    'Imputer', 'One Hot Encoder', 'Text Featurization Component', 'Elastic Net Classifier'])

# get error during pipeline fit
with pytest.raises(TypeError, match='invalid type promotion'):
    pipeline.fit(dt, y)
```

**Root Cause**
In our `TextFeaturizer`, we normalize to woodwork DataTable, get a pandas dataframe, run featuretools to compute some features, then run our LSA component to produce more features.

When we run the LSA component, we should be passing along the woodwork DataTable, because that contains the information that the "datetime" column happens to be NaturalLanguage-typed. However, instead we currently pass the pandas dataframe. This means the LSA component will run woodwork type inference, determine that "datetime" is a Datetime-typed feature, try to use that feature to generate TF-IDF features and error out.

[Here's the line](https://github.com/alteryx/evalml/blob/main/evalml/pipelines/components/transformers/preprocessing/text_featurizer.py#L122) which causes the problem.

**Fix**
Short-term: update that line to provide the woodwork DataTable instead of the pandas dataframe, to preserve the user-specified feature types.
Long-term: hey @freddyaboulton does your woodwork accessor feature branch fix this problem? If not, we should definitely make a fix on there as well.

### PR DESCRIPTION
**Pull request by dsherry**
Fix #2296 

Repro and root cause are on the issue.

There was more to the problem than I had originally thought. In addition to the text featurizer not passing along user-specified woodwork types to the LSA component (which is used inside the text featurizer), our component graph also wasn't preserving custom types during `fit`. I shored up our unit test coverage of this case a bit.

@freddyaboulton I bet the component graph changes will cause conflicts with your woodwork accessor feature branch. but hopefully they will be minor. I tried to write the tests in a way which meant you could avoid huge conflicts there at least.

### CODE REVIEW

File: evalml/pipelines/component_graph.py
Comment (dsherry) [inline evalml/pipelines/component_graph.py:123]: First bug: there's no need to convert to pandas here. That's all handled in `_compute_features`.

Symptom: calls to pipeline `fit` always use whatever types woodwork infers for all the columns, overriding all user-specified values.

Explanation: the first thing we do in `_compute_features` is call `X = infer_feature_types(X)`. So, by converting to pandas here and then immediately converting back to woodwork first thing in `_compute_features`, we were effectively resetting the input to whatever woodwork infers by default.

File: evalml/pipelines/components/transformers/preprocessing/text_featurizer.py
Comment (dsherry) [inline evalml/pipelines/components/transformers/preprocessing/text_featurizer.py:122]: This was the first fix required for this bug. We need to pass the woodwork type info to the LSA component, otherwise it will run woodwork inference again from scratch and the newly reset types will be used by all the downstream components!

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_component_graph.py:689]: No functional change, just reordered to make this test easier to understand

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_component_graph.py:703]: This change did not occur because of the reordering above; it was happening before. I think this changed because "3", "4" and "5" each have two occurrences, but the RF OHE has `top_n` set to 2, so the OHE has to choose two categories from those top three tied categories.

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_component_graph.py:742]: Woodwork would infer the data in "column_5" to be datetime, but I set it to natural language above. We check here that it passes through the datetime featurizer untouched and enters the text featurizer. On the next assert, we check that it gets transformed into natural language features.

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_component_graph.py:722]: The confusing thing about this test is that it was passing before my changes. That's weird because here we set an integer column "column_2" to have categorical type, and saw that as expected one-hot encoded features based off of "column_2" show up in the datetime component's input.

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_component_graph.py:372]: This isn't a "bug" per se, but the imputer will return one column for each input column, not more which is what this was doing

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_component_graph.py:378]: If you look at `X_y_binary`, all 20 columns in `X` have type `float64`.

File: evalml/pipelines/component_graph.py
Comment (dsherry) [inline evalml/pipelines/component_graph.py:331]: This method `_consolidate_inputs` currently only gets called inside `_compute_features`. If you look at the call site, you'll see the `x_inputs` is always a list of pandas dataframes. Therefore, after concatenating those pandas dataframes representing input features from the parent components in the graph, we need to call `_retain_custom_types_and_initalize_woodwork` to preserve any columns which appeared in the original woodwork input.

Comment (dsherry) [reply evalml/pipelines/component_graph.py:331]: I added `test_component_graph_types_multi_input_1` to check this is working properly.

File: evalml/pipelines/component_graph.py
Comment (dsherry) [inline evalml/pipelines/component_graph.py:124]: This, and the similar change in `_compute_features` are important. Otherwise, if the input provided to `fit` is not a woodwork structure, fit will fail!

File: evalml/pipelines/component_graph.py
Comment (dsherry) [inline evalml/pipelines/component_graph.py:325]: Since we're now expecting `_compute_features` will be normalizing input to woodwork up the stack, we need to convert to pandas here. You can see that this X input is guaranteed to be woodwork by the code in `_compute_features`.

File: evalml/tests/automl_tests/test_dask_engine.py
Comment (dsherry) [inline evalml/tests/automl_tests/test_dask_engine.py:157]: This test was flaking on me in the 17th decimal place, but only for [linux](https://github.com/alteryx/evalml/pull/2297/checks?check_run_id=2635860715)/[windows](https://github.com/alteryx/evalml/pull/2297/checks?check_run_id=2635667949) python 3.8 😆  not sure why exactly this PR would trigger that.

File: evalml/pipelines/component_graph.py
Comment (chukarsten) [reply evalml/pipelines/component_graph.py:123]: @dsherry this is very interesting and I wonder if it's part of what I was seeing in my investigation with Chris.

File: evalml/pipelines/component_graph.py
Comment (chukarsten) [reply evalml/pipelines/component_graph.py:325]: Ah, woodwork conversions

File: evalml/tests/component_tests/test_text_featurizer.py
Comment (chukarsten) [inline evalml/tests/component_tests/test_text_featurizer.py:238]: Are you able to explicitly just check the logical type of "col_2"?  or do an ordered match?  I think it would just be nice to see that the specific user specified type and column were the same before and after.  It would be a little clearer to test readers that need to go through and realize that all the other fields will be doubles.

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (chukarsten) [inline evalml/tests/pipeline_tests/test_component_graph.py:789]: So is this a smoke test for the estimator when fit() is called?  Where is the comment-mentioned test for column_3?

File: evalml/pipelines/component_graph.py
Comment (dsherry) [reply evalml/pipelines/component_graph.py:325]: Haha yep. This is necessary because there's no concatenate feature in woodwork `0.0.11`. I'm pretty sure this is handled on @freddyaboulton 's accessor feature branch

File: evalml/tests/component_tests/test_text_featurizer.py
Comment (dsherry) [reply evalml/tests/component_tests/test_text_featurizer.py:238]: Oh yeah, that's true.
So, the fact that "col_2" did not end up being converted into text features implies that the text featurizer saw it as categorical-type, not as natlang type.
But you're right, the extra step would be to do an ordered match. I'll do it.

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (dsherry) [reply evalml/tests/pipeline_tests/test_component_graph.py:789]: Lol yeah this is a little confusing. The assert a couple lines down checks that column_3 was provided as an input to the RF estimator. Before the impl changes in this PR, this test would fail during `fit`, because the code where we concatenate the various inputs would fall over.

I'll see if I can find a clean way to have this test focus on the types which each component sees, rather than on the feature names. That would definitely make the value of this test easier to follow.

File: evalml/pipelines/component_graph.py
Comment (freddyaboulton) [reply evalml/pipelines/component_graph.py:325]: Yep, in the ww update branch we just do `return_x = X` hehe but fyi there still isn't a concatenate feature in ww

**Comment by dsherry** [comment]
For those following along, this is still ready for review, but I'm waiting for #2181 to merge (woodwork accessor), then I'll rebase this and get it in.

**Review by bchen1116 — APPROVED** [review]
Nice! The tests look pretty thorough!

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (dsherry) [reply evalml/tests/pipeline_tests/test_component_graph.py:789]: Ok I think I got it

File: evalml/pipelines/component_graph.py
Comment (dsherry) [reply evalml/pipelines/component_graph.py:326]: Thanks for this @freddyaboulton . I'll try to file something for this before we merge this PR. Its a great point, and yep something we'll run into eventually.

File: evalml/tests/model_understanding_tests/test_partial_dependence.py
Comment (freddyaboulton) [inline evalml/tests/model_understanding_tests/test_partial_dependence.py:258]: What happened here is that since we updated the `ComponentGraph` to preserve the woodwork types throughout, card_id, store_id, customer_present get preserved as ints/bool so the predictions are different which makes the partial dependence results different.

This is a snapshot from the debugger of the ww types as soon as they make it to estimator.fit. Note that the ohe-columns are double instead of bool because of the standard scaler. 

### Main
woodwork_types_estimator_main


### This branch
image

File: evalml/tests/model_understanding_tests/test_partial_dependence.py
Comment (freddyaboulton) [inline evalml/tests/model_understanding_tests/test_partial_dependence.py:366]: I had posted a [previous comment](https://github.com/alteryx/evalml/pull/2297#discussion_r645096770) explaining why this change is necessary but it got wiped from the diff view with the black pr. 

In summary, since we're preserving types we were previously not preserving (bool/int) the predicted probabilities are different which means the partial dependence is different.

**Comment by freddyaboulton** [comment]
Perf tests [here](https://ayx.atlassian.net/wiki/spaces/AutoML/pages/513736705/Issue+2296+Preserve+Woodwork+Types+in+Component+Graph) 

Our linear estimators are doing worse so I think we should hold off on merging until we know why.

File: evalml/pipelines/component_graph.py
Comment (freddyaboulton) [inline evalml/pipelines/component_graph.py:397]: Before we would do this before the for-loop which was causing the performance regression in linear estimators

File: evalml/tests/pipeline_tests/test_component_graph.py
Comment (freddyaboulton) [inline evalml/tests/pipeline_tests/test_component_graph.py:1226]: Adding this to make sure we don't overwrite the types created by the standard scaler.

File: evalml/tests/model_understanding_tests/test_partial_dependence.py
Comment (freddyaboulton) [reply evalml/tests/model_understanding_tests/test_partial_dependence.py]: Disregard this and the previous comment! This was actually a symptom of a bug hehe

**Comment by freddyaboulton** [comment]
I've updated the performance tests [here](https://ayx.atlassian.net/wiki/spaces/AutoML/pages/513736705/Issue+2296+Preserve+Woodwork+Types+in+Component+Graph#Update-on-linear-estimators)! 

I was able to trace the issue to a bug where we were overriding the types set by the standard scaler. I updated the implementation so this should be good for review now!

**Comment by freddyaboulton** [comment]
@dsherry I can't add you as a reviewer 😂  But your feedback would be appreciated!

File: evalml/pipelines/component_graph.py
Comment (dsherry) [reply evalml/pipelines/component_graph.py:394]: I wonder if woodwork could help manage this for us. I guess adding a `concatenate` ability would do it.

Comment (freddyaboulton) [reply evalml/pipelines/component_graph.py:394]: @dsherry I think you're right that ww.concat would definitely make this easier! Looks like it'll be in the next release so I'm really excited to try it out!

https://github.com/alteryx/woodwork/pull/932

File: evalml/pipelines/components/transformers/preprocessing/text_featurizer.py
Comment (dsherry) [reply evalml/pipelines/components/transformers/preprocessing/text_featurizer.py:129]: Ohhhh. @freddyaboulton explained offline. Woodwork schema won't be copied by default pandas indexing operator, have to use woodwork namespace. Thanks @freddyaboulton !

**Pull request merged**
